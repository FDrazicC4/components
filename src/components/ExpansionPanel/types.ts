import { StyledComponent } from '@components/types';
import React, { ChangeEvent } from 'react';


export interface Props extends StyledComponent {


    /**
     * The content of the actions wrapper
     */
    actions?: React.ReactNode,


    /**
     * The content of the expandable wrapper
     */
    details: React.ReactNode,


    /**
     * Content used to display as the expand indicator
     */
    expandIcon?: React.ReactElement,


    /**
     * Determines if the default state is open or closed
     * @default 'false'
     */
    isDefaultExpanded?: boolean,


    /**
     * Determines if the panel is disabled or not
     * @default 'false'
     */
    isDisabled?: boolean,


    /**
     * If true, expands the panel. Collapses the panel if false.
     * This enables control over the panel.
     * @default 'false'
     */
    isExpanded?: boolean,


    /**
     * The callback fired when a panel is opened or closed
     */
    onChange?: (event: ChangeEvent<{}>, expanded: boolean) => void,


    /**
     * The content to display at all times
     */
    summary: React.ReactNode,


}
