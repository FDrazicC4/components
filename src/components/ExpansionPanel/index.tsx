import React from 'react';
import { ExpansionPanel as MuiExpansionPanel } from '@material-ui/core';
import { ExpansionPanelSummary as MuiExpansionPanelSummary } from '@material-ui/core';
import { ExpansionPanelDetails as MuiExpansionPanelDetails } from '@material-ui/core';
import { ExpansionPanelActions as MuiExpansionPanelActions } from '@material-ui/core';
import { Props } from './types';


/**
 * Expansion panels contain a summary of content that can be collapsible
 * Can be used as a standalone component or combined with other components like cards
 */
export const ExpansionPanel = React.forwardRef(({
    actions,
    details,
    expandIcon,
    hasDefaultExpanded,
    isDisabled,
    isExpanded,
    isSquare,
    onChange,
    summary,
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => {


    const wrapperClasses = {
        root: 'brew-ExpansionPanel',
        rounded: 'brew-ExpansionPanel--isRounded',
        expanded: 'brew-ExpansionPanel--isExpanded',
        disabled: 'brew-ExpansionPanel--isDisabled',
    };


    const summaryClasses = {
        root: 'brew-ExpansionPanel__summary',
        content: 'brew-ExpansionPanel__summaryContent',
        expandIcon: 'brew-ExpansionPanel__expandIcon',
    };


    return (
        <MuiExpansionPanel
            classes={wrapperClasses}
            defaultExpanded={hasDefaultExpanded}
            disabled={isDisabled}
            expanded={isExpanded}
            onChange={onChange}
            ref={ref}
            square={isSquare}
            {...otherProps}
        >
            <MuiExpansionPanelSummary classes={summaryClasses} expandIcon={expandIcon}>
                {summary}
            </MuiExpansionPanelSummary>
            <MuiExpansionPanelDetails className="brew-ExpansionPanel__details">
                {details}
            </MuiExpansionPanelDetails>
            {actions && (
                <MuiExpansionPanelActions className="brew-ExpansionPanel__actions">
                    {actions}
                </MuiExpansionPanelActions>
            )}
        </MuiExpansionPanel>
    );


});


ExpansionPanel.displayName = 'ExpansionPanel';


export default ExpansionPanel;
