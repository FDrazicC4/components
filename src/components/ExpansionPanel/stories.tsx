import React from 'react';
import { boolean, text } from '@storybook/addon-knobs';
import ExpansionPanel from '@components/ExpansionPanel';
import Flag from '@components/Flag';
import Icon from '@components/Icon';
import Input from '@components/Input';
import Button from '@components/Button';


export default {
    component: ExpansionPanel,
    parameters: {
        componentSubtitle: <Flag color="success">Stable</Flag>,
    },
    title: 'Stable|Data Display/ExpansionPanel',
}


export const Sandbox = (): React.ReactElement => (
        <div>
            <ExpansionPanel
                summary="Brief Summary"
                details="Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                an unknown printer took a galley of type and scrambled it to make a type specimen book.
                It has survived not only five centuries, but also the leap into electronic typesetting,
                remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset
                sheets containing Lorem Ipsum passages, and more recently with desktop publishing software
                like Aldus PageMaker including versions of Lorem Ipsum"
            />
            <ExpansionPanel
                details="Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                an unknown printer took a galley of type and scrambled it to make a type specimen book.
                It has survived not only five centuries, but also the leap into electronic typesetting,
                remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset
                sheets containing Lorem Ipsum passages, and more recently with desktop publishing software
                like Aldus PageMaker including versions of Lorem Ipsum"
                expandIcon={<Icon>arrow_drop_up</Icon>}
                isDefaultExpanded
                summary="Set the default state to expanded"
            />
            <ExpansionPanel
                isDisabled
                details="Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                an unknown printer took a galley of type and scrambled it to make a type specimen book.
                It has survived not only five centuries, but also the leap into electronic typesetting,
                remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset
                sheets containing Lorem Ipsum passages, and more recently with desktop publishing software
                like Aldus PageMaker including versions of Lorem Ipsum"
                expandIcon={<Icon>arrow_drop_up</Icon>}
                summary="You may also disable a certain panel"
            />
        </div>
);


export const Controlled = (): React.ReactElement => {


    const [isExpanded, setIsExpanded] = React.useState<string | false>(false);


    const handleChange = (panel: string) => (event: React.ChangeEvent<{}>, expanded: boolean) => {
        setIsExpanded(expanded ? panel : false);
    };


    return (
        <div>
            <ExpansionPanel
                expandIcon={<Icon>arrow_drop_up</Icon>}
                summary="Brief Summary"
                details="Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                an unknown printer took a galley of type and scrambled it to make a type specimen book.
                It has survived not only five centuries, but also the leap into electronic typesetting,
                remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset
                sheets containing Lorem Ipsum passages, and more recently with desktop publishing software
                like Aldus PageMaker including versions of Lorem Ipsum"
                isExpanded={isExpanded === 'panel1'}
                onChange={handleChange('panel1')}
            />
            <ExpansionPanel
                details="Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                an unknown printer took a galley of type and scrambled it to make a type specimen book.
                It has survived not only five centuries, but also the leap into electronic typesetting,
                remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset
                sheets containing Lorem Ipsum passages, and more recently with desktop publishing software
                like Aldus PageMaker including versions of Lorem Ipsum"
                expandIcon={<Icon>arrow_drop_up</Icon>}
                isExpanded={isExpanded === 'panel2'}
                onChange={handleChange('panel2')}
                summary="Set the default state to expanded"
            />
            <ExpansionPanel
                details="Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                an unknown printer took a galley of type and scrambled it to make a type specimen book.
                It has survived not only five centuries, but also the leap into electronic typesetting,
                remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset
                sheets containing Lorem Ipsum passages, and more recently with desktop publishing software
                like Aldus PageMaker including versions of Lorem Ipsum"
                expandIcon={<Icon>arrow_drop_up</Icon>}
                isExpanded={isExpanded === 'panel3'}
                onChange={handleChange('panel3')}
                summary="You may also disable a certain panel"
            />
        </div>
    );


};


export const Custom = (): React.ReactElement => {


    return (
        <div>
            <ExpansionPanel
                expandIcon={<Icon>arrow_drop_up</Icon>}
                summary={<Input.Checkbox label="I acknowledge" />}
                details="Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                an unknown printer took a galley of type and scrambled it to make a type specimen book.
                It has survived not only five centuries, but also the leap into electronic typesetting,
                remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset
                sheets containing Lorem Ipsum passages, and more recently with desktop publishing software
                like Aldus PageMaker including versions of Lorem Ipsum"
                aria-label="panel"
            />
            <ExpansionPanel
                summary={
                    <div>
                        <Button size="small" variant="text" onClick={e => e.stopPropagation()}>Click Me</Button>
                        Stop Propagation of the click event to not expand when clicking the button.
                    </div>
                }
                details="Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                an unknown printer took a galley of type and scrambled it to make a type specimen book.
                It has survived not only five centuries, but also the leap into electronic typesetting,
                remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset
                sheets containing Lorem Ipsum passages, and more recently with desktop publishing software
                like Aldus PageMaker including versions of Lorem Ipsum"
                expandIcon={<Icon>arrow_drop_up</Icon>}
            />
            <ExpansionPanel
                details="Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                an unknown printer took a galley of type and scrambled it to make a type specimen book.
                It has survived not only five centuries, but also the leap into electronic typesetting,
                remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset
                sheets containing Lorem Ipsum passages, and more recently with desktop publishing software
                like Aldus PageMaker including versions of Lorem Ipsum"
                expandIcon={<Icon>arrow_drop_up</Icon>}
                isDisabled
                summary={<Input.Checkbox label="I must stop click propagation for this to work" />}
            />
        </div>
    );


};
