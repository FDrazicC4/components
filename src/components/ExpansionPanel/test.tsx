import React from 'react';
import renderer from 'react-test-renderer';
import ExpansionPanel from '@components/ExpansionPanel';


it('renders correctly', () => {
    const tree = renderer
        .create((
            <React.Fragment>
                <ExpansionPanel
                    details="bar"
                    isDisabled
                    summary="foo"
                />
                <ExpansionPanel
                    details="goodbye"
                    hasDefaultExpanded
                    summary="hello"
                />
            </React.Fragment>
        ))
        .toJSON();
    expect(tree).toMatchSnapshot();
});