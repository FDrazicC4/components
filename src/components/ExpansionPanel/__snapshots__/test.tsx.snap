// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="MuiPaper-root MuiExpansionPanel-root brew-ExpansionPanel Mui-disabled brew-ExpansionPanel--isDisabled MuiExpansionPanel-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled={true}
      aria-expanded={false}
      className="MuiButtonBase-root MuiExpansionPanelSummary-root brew-ExpansionPanel__summary Mui-disabled Mui-disabled"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={-1}
    >
      <div
        className="MuiExpansionPanelSummary-content brew-ExpansionPanel__summaryContent"
      >
        <div
          className="MuiTypography-root brew-Typography brew-Typography--display-initial brew-Typography--variant-body1 brew-Typography--align-inherit brew-Typography--color-inherit MuiTypography-body1"
        >
          foo
        </div>
      </div>
    </div>
    <div
      className="MuiCollapse-container MuiCollapse-hidden"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              className="MuiExpansionPanelDetails-root brew-ExpansionPanel__details"
            >
              <div
                className="MuiTypography-root brew-Typography brew-Typography--display-initial brew-Typography--variant-body1 brew-Typography--align-inherit brew-Typography--color-inherit MuiTypography-body1"
              >
                bar
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="MuiPaper-root MuiExpansionPanel-root brew-ExpansionPanel Mui-expanded brew-ExpansionPanel--isExpanded MuiExpansionPanel-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled={false}
      aria-expanded={true}
      className="MuiButtonBase-root MuiExpansionPanelSummary-root brew-ExpansionPanel__summary Mui-expanded"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiExpansionPanelSummary-content brew-ExpansionPanel__summaryContent Mui-expanded"
      >
        <div
          className="MuiTypography-root brew-Typography brew-Typography--display-initial brew-Typography--variant-body1 brew-Typography--align-inherit brew-Typography--color-inherit MuiTypography-body1"
        >
          hello
        </div>
      </div>
    </div>
    <div
      className="MuiCollapse-container MuiCollapse-entered"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              className="MuiExpansionPanelDetails-root brew-ExpansionPanel__details"
            >
              <div
                className="MuiTypography-root brew-Typography brew-Typography--display-initial brew-Typography--variant-body1 brew-Typography--align-inherit brew-Typography--color-inherit MuiTypography-body1"
              >
                goodbye
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
